### DISCUSSION TITLE
How to debug stuck VM / hang with no shell access?

### DISCUSSION THREAD

**Discussion by ian-h-chamberlain** [abiosoft/colima#422] [Q&A]
Hello, recently I have been facing an issue while using Colima that I'm kind of at a loss for how to debug. The issue results in a state where I can no longer access any containers and not even the colima VM, and my only recourse is to `colima stop --force` and start again.

Basically, something in my organization's build process (just a basic Makefile) causes the container and even the VM to lock up whenever I build certain targets with `make`. I cannot use `colima ssh`, nor any `docker` commands — they all seem to just hang forever when I try.

Unfortunately, I can't share the code as it's private to our organization, but I'm hoping someone here might be able to help me get started on debugging this. I've tried adding different kinds of tracing to the build to see if I can figure out what specific command causes it, but so far I've had little luck.

Is there any kind of "backdoor" way into the VM or its filesystem that I might be able to use to help with this? At this point I'm just kind of stuck because I can't build anything using `colima` and I'm afraid I might have to switch to another solution for this use case, but I have had a great experience using it for other purposes so I'm hoping I can figure out how to work around or fix this. Thanks in advance!

**Comment by SuperFola** [comment]
Could it be because you are building huge executables and it takes all the CPU time allocated + RAM allocated to the VM, making it unresponsive?

Have you tried giving more RAM and CPU to the VM?

**Reply by ian-h-chamberlain** [reply]
Wow, maybe that was it! I thought I had done this once before already, but I increased it even more (8 core + 8G memory) just to be certain and it seems like I am able to build now... I wouldn't have guessed this particular build to be so resource-intensive, but this simple suggestion seems to have gotten me unstuck. Thanks!

**Reply by ian-h-chamberlain** [reply]
Hmm, maybe I spoke too soon. A few hours after I posted this I encountered the freeze again, so maybe I just got lucky the first time. It seems that perhaps only some build targets cause the problem, so there might be a specific command somewhere in there causing it.

**Reply by abiosoft** [reply]
From my experience, this happens when the VM run out of resources (usually memory). 
`colima stop -f` can be used to forcefully shut it down.

I am not sure you would be able to actually debug as the primary interaction medium(ssh) becomes unresponsive as well.

**Reply by ian-h-chamberlain** [reply]
Revisiting this as I ran into it again today, and managed to reproduce while running `docker stats`...

You can see the output cut off once the VM became unresponsive. These numbers don't seem particularly horrible, considering the VM had 4 CPUs allotted... 

```
CONTAINER ID   NAME               CPU %     MEM USAGE / LIMIT     MEM %     NET I/O        BLOCK I/O       PIDS
4779296d1270   flamboyant_haibt   296.12%   1.169GiB / 7.776GiB   15.04%    26kB / 3.1kB   498MB / 197kB   137
CONTAINER ID   NAME               CPU %     MEM USAGE / LIMIT     MEM %     NET I/O        BLOCK I/O       PIDS
4779296d1270   flamboyant_haibt   296.12%   1.169GiB / 7.776GiB   15.04%    26kB / 3.1kB   498MB / 197kB   137
CONTAINER ID   NAME               CPU %     MEM USAGE / LIMIT     MEM %     NET I/O        BLOCK I/O       PIDS
4779296d1270   flamboyant_haibt   296.12%   1.169GiB / 7.776GiB   15.04%    26kB / 3.1kB   498MB / 197kB   137
CONTAINER ID   NAME               CPU %     MEM USAGE / LIMIT     MEM %     NET I/O        BLOCK I/O       PIDS
4779296d1270   flamboyant_haibt   296.12%   1.169GiB / 7.776GiB   15.04%    26kB / 3.1kB   498MB / 197kB   137
CONTAINER ID   NAME               CPU %     MEM USAGE / LIMIT   MEM %     NET I/O   BLOCK I/O   PIDS
4779296d1270   flamboyant_haibt   --        -- / --             --        --        --          --
CONTAINER ID   NAME               CPU %     MEM USAGE / LIMIT   MEM %     NET I/O   BLOCK I/O   PIDS
4779296d1270   flamboyant_haibt   --        -- / --             --        --        --          --
CONTAINER ID   NAME               CPU %     MEM USAGE / LIMIT   MEM %     NET I/O   BLOCK I/O   PIDS
4779296d1270   flamboyant_haibt   --        -- / --             --        --        --          --
CONTAINER ID   NAME               CPU %     MEM USAGE / LIMIT   MEM %     NET I/O   BLOCK I/O   PIDS
```

Any other ideas for stuff I could look at to figure out why this might be happening?

**Reply by SuperFola** [reply]
CPU usage is very high, about 3/4 of what you allocated. This could be a problem with the docker daemon inside the VM, and not just a "badly built VM".
If you don't destroy the VM and recreate it when this happens, you could have a look at the system logs and look for docker related events/commands/etc?

**Reply by ian-h-chamberlain** [reply]
> you could have a look at the system logs and look for docker related events/commands/etc?

Do you mean my host system? `colima status`, `ssh` and any `docker` commands all seem to become unresponsive in this scenario so I guess I'm not sure where else to look.

Regarding high CPU — yes, it is a lot, but I think it's to be expected to some extent while running a build. Normally while idle the container runs from 1—25% but of course compiling code takes significantly more resources. I have seen the percentage creep up well into the 300%+ range without "crashing" like this, so I don't _think_ it's hitting some specific threshold, and in fact sometimes the whole build actually succeeds.

Let me see if there are any other processes running in the container that might be complicating this, and try to pin it down a bit further.

**Reply by ian-h-chamberlain** [reply]
One more data point — I eliminated some extraneous stuff in the container and reproduced again, this time with very little resource usage!
```
CONTAINER ID   NAME               CPU %     MEM USAGE / LIMIT     MEM %     NET I/O      BLOCK I/O        PIDS
4779296d1270   flamboyant_haibt   33.33%    21.39MiB / 7.776GiB   0.27%     1.3kB / 0B   1.41MB / 102kB   8
CONTAINER ID   NAME               CPU %     MEM USAGE / LIMIT     MEM %     NET I/O      BLOCK I/O        PIDS
4779296d1270   flamboyant_haibt   33.33%    21.39MiB / 7.776GiB   0.27%     1.3kB / 0B   1.41MB / 102kB   8
CONTAINER ID   NAME               CPU %     MEM USAGE / LIMIT   MEM %     NET I/O   BLOCK I/O   PIDS
4779296d1270   flamboyant_haibt   --        -- / --             --        --        --          --
```

**Reply by SuperFola** [reply]
> Do you mean my host system? 

I meant in the VM, this is a full blown linux system, there should be some logs

**Reply by abiosoft** [reply]
Are you emulating x86_64 on M1 or vice-versa?

**Reply by ian-h-chamberlain** [reply]
No, I'm on an x86_64 host with an x86_64 VM. Since any `colima` command seems to lock up when this happens, is there some other way (maybe some magic `limactl` commands) or places on the host I can look for logs if I reproduce again?
